eth/executionclient/execution_client.go

```diff
@@ -31,76 +32,164 @@ var (
 	connectionTimeout           time.Duration
 	reconnectionInitialInterval time.Duration
 	reconnectionMaxInterval     time.Duration
 	logBatchSize                uint64
 
 	syncDistanceTolerance uint64
-	syncProgressFn        func(context.Context) (*ethereum.SyncProgress, error)
+	syncProgressFn        func(context.Context, *ManagedClient) (*ethereum.SyncProgress, error)
 
 	// variables
-	client *ethclient.Client
-	closed chan struct{}
+	clientsMu   sync.RWMutex
+	clients     []*ManagedClient
+	closed      chan struct{}
+	closingOnce sync.Once
 }
 
 // New creates a new instance of ExecutionClient.
 func New(ctx context.Context, nodeAddr string, contractAddr ethcommon.Address, opts ...Option) (*ExecutionClient, error) {
+	addrList := strings.Split(nodeAddr, ";")
+	if len(addrList) == 0 {
+		return nil, fmt.Errorf("no node address provided")
+	}
+
 	client := &ExecutionClient{
-		nodeAddr:                    nodeAddr,
+		nodeAddrs:                   addrList,
 		contractAddress:             contractAddr,
 		logger:                      zap.NewNop(),
 		followDistance:              DefaultFollowDistance,
 		connectionTimeout:           DefaultConnectionTimeout,
 		reconnectionInitialInterval: DefaultReconnectionInitialInterval,
 		reconnectionMaxInterval:     DefaultReconnectionMaxInterval,
-		logBatchSize:                DefaultHistoricalLogsBatchSize, // TODO Make batch of logs adaptive depending on "websocket: read limit"
+		logBatchSize:                DefaultHistoricalLogsBatchSize,
 		closed:                      make(chan struct{}),
 	}
+
 	for _, opt := range opts {
 		opt(client)
 	}
-	err := client.connect(ctx)
+
+	// Initialize ManagedClients
+	for _, addr := range client.nodeAddrs {
+		mc, err := NewManagedClient(ctx, addr, client.logger, client.reconnectionInitialInterval, client.reconnectionMaxInterval, client.syncDistanceTolerance)
+		if err != nil {
+			client.logger.Error("Failed to initialize ManagedClient", zap.String("address", addr), zap.Error(err))
+			return nil, fmt.Errorf("create managed client: %w", err)
+		}
+		client.clients = append(client.clients, mc)
+	}
+
+	if len(client.clients) == 0 {
+		return nil, fmt.Errorf("no execution clients could be initialized")
+	}
+
+	same, err := client.assertSameChainIDs(ctx)
 	if err != nil {
-		return nil, fmt.Errorf("failed to connect to execution client: %w", err)
+		return nil, fmt.Errorf("assert same chain IDs: %w", err)
+	}
+	if !same {
+		return nil, fmt.Errorf("execution clients' chain IDs are not same")
 	}
 
 	client.syncProgressFn = client.syncProgress
 
 	return client, nil
 }
 
-func (ec *ExecutionClient) syncProgress(ctx context.Context) (*ethereum.SyncProgress, error) {
-	return ec.client.SyncProgress(ctx)
+// assertSameChainIDs checks if all healthy clients have the same chain ID.
+// It sets firstChainID to the chain ID of the first healthy client encountered.
+func (ec *ExecutionClient) assertSameChainIDs(ctx context.Context) (bool, error) {
```

Comment:
since this returns err, it's recommended to propagate the error message here then print on a higher level. doing both can result in a lot of redundancy in the logs.